### PR TITLE
#472_46 - Add image switching functionality

### PR DESCRIPTION
**Pull request by ntoombs19**
- Remove README file
- Change registration.php to fix mispelling
- Add InstallData script to include new image attribute
- Add css for image switching transition
- Add plugin to override xsd
- Add preference to include new alt_iage url
- Add logic to template to display alt image

### CODE REVIEW

File: Block/Product/ImageBuilder.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
+<?php
+/**
+ * @category    ClassyLlama
```
Comment (erikhansen) [inline Block/Product/ImageBuilder.php]: Change the copyright block to match this one: https://github.com/classyllama/ClassyLlama_AvaTax/blob/develop/Block/CustomerAddress.php#L1..L13 Make this change to all files in this extension.

Also, add a LICENSE.txt file like this to the root of the repo: https://github.com/classyllama/ClassyLlama_AvaTax/blob/develop/LICENSE.txt

File: Block/Product/ImageBuilder.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                $this->product,
+                'category_page_alt',
+                [
+                    'width' => $helper->getWidth(),
+                    'height' => $helper->getHeight()
+                ]
+            );
+
+        $altImageUrl = $altImageHelper->getUrl();
+        $altImageUrl = strpos($altImageUrl, '/placeholder/') ? false : $altImageUrl;
```
Comment (erikhansen) [inline Block/Product/ImageBuilder.php]: This seems like a janky way to check for whether the alternative product image exists. Can you come up with a cleaner way to check this?

File: Plugin/Framework/Config/SchemaLocator.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
+<?php
+/**
+ * @category    ClassyLlama
+ * @copyright   Copyright (c) 2017 Classy Llama
+ */
+
+namespace MagentoEse\ProductImageSwitcher\Plugin\Framework\Config;
+
+class SchemaLocator {
```
Comment (erikhansen) [inline Plugin/Framework/Config/SchemaLocator.php]: Drop curly brace to a new line

File: Setup/InstallData.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+use Magento\Catalog\Model\ResourceModel\Eav\Attribute;
+
+class InstallData implements \Magento\Framework\Setup\InstallDataInterface
+{
+    /**
+     * EAV setup factory
+     *
+     * @var eavSetupFactory
+     */
+    private $eavSetupFactory;
```
Comment (erikhansen) [inline Setup/InstallData.php]: Change this to "protected"

File: Setup/InstallData.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * @param \Magento\Framework\Setup\ModuleDataSetupInterface $setup
+     * @param \Magento\Framework\Setup\ModuleContextInterface $context
+     * @throws \Exception
+     */
+    public function install(
+        \Magento\Framework\Setup\ModuleDataSetupInterface $setup,
+        \Magento\Framework\Setup\ModuleContextInterface $context)
```
Comment (erikhansen) [inline Setup/InstallData.php]: Change formatting to:
```
        \Magento\Framework\Setup\ModuleContextInterface $context
    ) {
```

File: Setup/InstallData.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+     * InstallData constructor.
+     * @param \Magento\Eav\Setup\EavSetupFactory $eavSetupFactory
+     */
+    public function __construct(\Magento\Eav\Setup\EavSetupFactory $eavSetupFactory)
+    {
+        $this->eavSetupFactory = $eavSetupFactory;
+    }
+
+    /**
+     * @param \Magento\Framework\Setup\ModuleDataSetupInterface $setup
```
Comment (erikhansen) [inline Setup/InstallData.php]: Add PhpDoc description (can be auto-generated)

File: etc/view.xsd [hunk 7]
```diff
@@ -0,0 +1,129 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+/**
+ * @category    ClassyLlama
+ * @copyright   Copyright (c) 2017 Classy Llama
+ */
+-->
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
```
Comment (erikhansen) [inline etc/view.xsd]: Add a comment to the top of this file with:

1. Explanation of why this file had to be copied forward. Maybe just point the reader to read the comment on `MagentoEse\ProductImageSwitcher\Plugin\Framework\Config::afterGetSchema`

2. Make note of exactly what has been changed in this file, maybe with BEGIN/END edit comments.

File: view/frontend/web/css/source/_module.less [hunk 8]
```diff
@@ -0,0 +1,13 @@
+/**
+ * @category    ClassyLlama
+ * @copyright   Copyright (c) 2017 Classy Llama
+ */
+
+& when (@media-common = true) {
+    .product-image-wrapper .product-image-photo:not(.no-alt-image) {
+        .lib-css(transition, opacity 300ms ease-in-out);
+        &:not(.alt-image):hover {
```
Comment (erikhansen) [inline view/frontend/web/css/source/_module.less]: Can you test the "hover" behavior on an iOS simulator? I want to make sure the behavior is not odd on a touch device. For example, what happens when you "touch" a PLP image on a touch device? Does the "alt" image flash right before you are taken to the PDP? Maybe we can review this together?

File: view/frontend/templates/product/image_with_borders.phtml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+/** @var $block \Magento\Catalog\Block\Product\Image */
+
+$altImageUrl = $block->getAltImageUrl();
+?>
+
+<span class="product-image-container"
+      style="width:<?= $block->getWidth()?>px;">
+    <span class="product-image-wrapper"
+          style="padding-bottom: <?= ($block->getRatio() * 100); ?>%;">
+        <?php if($altImageUrl): ?>
```
Comment (erikhansen) [inline view/frontend/templates/product/image_with_borders.phtml]: Add space after "if"

File: view/frontend/web/css/source/_module.less [hunk 10]
```diff
@@ -0,0 +1,13 @@
+/**
+ * @category    ClassyLlama
+ * @copyright   Copyright (c) 2017 Classy Llama
+ */
+
+& when (@media-common = true) {
+    .product-image-wrapper .product-image-photo:not(.no-alt-image) {
```
Comment (erikhansen) [inline view/frontend/web/css/source/_module.less]: From just reading this CSS, I have a hard time imagining how it's actually working, but I guess I'll just have to see it in practice.

File: Block/Product/ImageBuilder.php (hunk 2 above)
Comment (ntoombs19) [inline Block/Product/ImageBuilder.php]: It is admittedly janky but there isn't a cleaner way without writing a preference to override the getUrl method. "/placeholder/" is hard coded into the path for placeholder images so it should work consistently.

File: view/frontend/web/css/source/_module.less (hunk 10 above)
Comment (ntoombs19) [inline view/frontend/web/css/source/_module.less]: If a product doesn't have an alternate image, it gets the class 'no-alt-image' added to it. This prevents the hover styles being applied to that element when there is no alt image to display.

**Pull request merged**